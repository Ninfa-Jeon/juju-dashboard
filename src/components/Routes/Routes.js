import { useEffect } from "react";
import { Route, Redirect, Switch, useLocation } from "react-router-dom";

import Login from "components/LogIn/LogIn";

import Controllers from "pages/Controllers/Controllers";
import Models from "pages/Models/Models";
import ModelDetails from "pages/ModelDetails/ModelDetails";
import Settings from "pages/Settings/Settings";
import NotFound from "pages/NotFound/NotFound";

import useAnalytics from "hooks/useAnalytics";

export const paths = {
  "/": { redirect: "/models" },
  "/models": { component: Models },
  "/models/*": { component: ModelDetails },
  "/controllers/": { component: Controllers },
  "/settings": { component: Settings },
};

export function Routes() {
  const sendAnalytics = useAnalytics();
  const location = useLocation();

  useEffect(() => {
    // Send an analytics event when the URL changes.
    sendAnalytics({
      path: window.location.href.replace(window.location.origin, ""),
    });
  }, [location, sendAnalytics]);

  const routes = Object.entries(paths).map((path) => {
    const Component = path[1].component;
    if (path[1].redirect) {
      return (
        <Route key={path[0]} path={path[0]} exact>
          <Redirect to={path[1].redirect} />
        </Route>
      );
    }
    return (
      <Route key={path[0]} path={path[0]} exact>
        <Login>
          <Component />
        </Login>
      </Route>
    );
  });

  return (
    <Switch>
      {routes}
      <Route>
        <NotFound />
      </Route>
    </Switch>
  );
}
