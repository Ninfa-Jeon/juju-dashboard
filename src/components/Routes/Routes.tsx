import { useEffect } from "react";
import { Route, Redirect, Switch, useLocation } from "react-router-dom";

import Login from "components/LogIn/LogIn";

import Controllers from "pages/Controllers/Controllers";
import Models from "pages/Models/Models";

// Entity Detail pages
import Model from "pages/EntityDetails/Model/Model";
import App from "pages/EntityDetails/App/App";

import Settings from "pages/Settings/Settings";
import NotFound from "pages/NotFound/NotFound";

import useAnalytics from "hooks/useAnalytics";

type Path = {
  redirect?: string;
  component?: () => JSX.Element;
};

type Paths = {
  [key: string]: Path;
};

export type EntityDetailsRoute = {
  userName: string;
  modelName: string;
};

export const paths: Paths = {
  "/": { redirect: "/models" },
  "/models": { component: Models },
  "/models/:userName/:modelName?": { component: Model },
  "/models/:userName/:modelName?/app/:appName?": { component: App },
  "/controllers": { component: Controllers },
  "/settings": { component: Settings },
};

export function Routes() {
  const sendAnalytics = useAnalytics();
  const location = useLocation();

  useEffect(() => {
    // Send an analytics event when the URL changes.
    sendAnalytics({
      path: window.location.href.replace(window.location.origin, ""),
    });
  }, [location, sendAnalytics]);

  const routes = Object.entries(paths).map((path) => {
    const Component = path[1].component;
    if (path[1].redirect) {
      return (
        <Route key={path[0]} path={path[0]} exact>
          <Redirect to={path[1].redirect} />
        </Route>
      );
    }
    return (
      <Route key={path[0]} path={path[0]} exact>
        <Login>{Component ? <Component /> : null}</Login>
      </Route>
    );
  });

  return (
    <Switch>
      {routes}
      <Route>
        <NotFound />
      </Route>
    </Switch>
  );
}
