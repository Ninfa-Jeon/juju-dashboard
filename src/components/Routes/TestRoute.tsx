import { Route } from "react-router-dom";

type Props = {
  path: string;
  children: JSX.Element;
};

export default function TestRoute({ path, children }: Props) {
  return <Route path={path}>{children}</Route>;
}
