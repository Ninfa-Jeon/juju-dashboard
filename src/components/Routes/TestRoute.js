import React from "react";
import { Route } from "react-router-dom";

import { paths } from "./Routes";

export default function TestRoute({ path, children } = {}) {
  // Validate paths
  if (Object.keys(paths).indexOf(path) < 0) {
    throw new Error("Path not found in application routes");
  }
  return <Route path={path}>{children}</Route>;
}
