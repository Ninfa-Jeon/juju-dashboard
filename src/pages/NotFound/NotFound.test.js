import React from "react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import dataDump from "testing/complete-redux-store-dump";

import { Routes } from "components/Routes/Routes";

const mockStore = configureStore([]);

describe("NotFound page", () => {
  it("should display when unknown route is accessed", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/foobar11"]}>
          <Routes />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("NotFound").length).toBe(1);
    // Ensure only one route is rendered
    expect(wrapper.find("main").length).toBe(1);
  });
});
