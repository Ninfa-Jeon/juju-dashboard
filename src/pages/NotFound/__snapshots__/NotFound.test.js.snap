// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFound page renders without crashing and matches snapshot 1`] = `
<Layout>
  <div
    className="p-strip"
  >
    <div
      className="row"
    >
      <h2>
        ¯\\_(ツ)_/¯
      </h2>
    </div>
  </div>
</Layout>
`;
